File: src/core/timeline/edit_controls_manager.py
"""Edit controls manager for timeline editing operations."""

from typing import List, Optional, Dict, Any, Set
from uuid import UUID
import logging

from .models import TimelineSegment, TimelineTrack
from .state import TimelineStateManager
from .edit_operations import EditOperations

logger = logging.getLogger(__name__)


class EditControlsManager:
    """Manager for timeline edit controls and operations."""

    def __init__(self, state_manager: TimelineStateManager):
        self.state_manager = state_manager
        self.edit_operations = EditOperations(state_manager)
        self.selected_segments: Set[UUID] = set()
        self.selected_tracks: Set[UUID] = set()

    def select_segment(self, segment_id: UUID) -> None:
        """Select a single segment."""
        self.selected_segments.clear()
        self.selected_tracks.clear()
        self.selected_segments.add(segment_id)
        self.state_manager.select_segment(segment_id)
        logger.info(f"Selected segment {segment_id}")

    def select_track(self, track_id: UUID) -> None:
        """Select a track and all its segments."""
        self.selected_tracks.clear()
        self.selected_segments.clear()
        self.selected_tracks.add(track_id)
        self.state_manager.select_track(track_id)
        logger.info(f"Selected track {track_id}")

    def select_time_range(self, start_time: float, end_time: float) -> None:
        """Select all segments in a time range."""
        self.selected_segments.clear()
        self.selected_tracks.clear()
        # Implementation would find segments in time range
        self.state_manager.selection_state.selection_start_time = start_time
        self.state_manager.selection_state.selection_end_time = end_time
        logger.info(f"Selected time range {start_time}-{end_time}")

    def clear_selection(self) -> None:
        """Clear all selections."""
        self.selected_segments.clear()
        self.selected_tracks.clear()
        self.state_manager.clear_selection()
        logger.info("Cleared all selections")

    def cut_at_time(self, cut_time: float) -> List[TimelineSegment]:
        """Cut selected segments at the specified time."""
        if not self.selected_segments:
            logger.warning("No segments selected for cutting")
            return []

        new_segments = self.edit_operations.cut_segments(
            list(self.selected_segments),
            cut_time
        )

        # Update selection to new segments
        self.selected_segments.clear()
        self.selected_segments.update(s.id for s in new_segments)

        return new_segments

    def copy_selection(self) -> List[TimelineSegment]:
        """Copy selected segments to clipboard."""
        if not self.selected_segments:
            logger.warning("No segments selected for copying")
            return []

        return self.edit_operations.copy_segments(list(self.selected_segments))

    def paste_at_time(self, paste_time: float, target_track_id: Optional[UUID] = None) -> List[TimelineSegment]:
        """Paste segments from clipboard at specified time."""
        return self.edit_operations.paste_segments(paste_time, target_track_id)

    def delete_selection(self) -> bool:
        """Delete selected segments."""
        if not self.selected_segments:
            logger.warning("No segments selected for deletion")
            return False

        success = self.edit_operations.delete_segments(list(self.selected_segments))
        if success:
            self.selected_segments.clear()

        return success

    def trim_selection(self, trim_start: bool = True, trim_end: bool = False,
                      trim_time: float = 0.0) -> Optional[TimelineSegment]:
        """Trim selected segment."""
        if len(self.selected_segments) != 1:
            logger.warning("Can only trim single segment selection")
            return None

        segment_id = list(self.selected_segments)[0]
        return self.edit_operations.trim_segment(segment_id, trim_start, trim_end, trim_time)

    def split_selection(self, split_time: float) -> Optional[tuple]:
        """Split selected segment at specified time."""
        if len(self.selected_segments) != 1:
            logger.warning("Can only split single segment selection")
            return None

        segment_id = list(self.selected_segments)[0]
        return self.edit_operations.split_segment(segment_id, split_time)

    def undo_last_operation(self) -> bool:
        """Undo the last edit operation."""
        if not self.state_manager.can_undo():
            logger.warning("No operations to undo")
            return False

        state_data = self.state_manager.undo()
        if state_data:
            logger.info("Undid last operation")
            return True

        return False

    def redo_last_operation(self) -> bool:
        """Redo the last undone operation."""
        if not self.state_manager.can_redo():
            logger.warning("No operations to redo")
            return False

        state_data = self.state_manager.redo()
        if state_data:
            logger.info("Redid last operation")
            return True

        return False

    def get_clipboard_info(self) -> Dict[str, Any]:
        """Get information about clipboard contents."""
        return {
            "segment_count": len(self.edit_operations.clipboard),
            "total_duration": sum(
                s.end_time - s.start_time
                for s in self.edit_operations.clipboard
            ),
            "segments": [
                {
                    "id": str(s.id),
                    "start_time": s.start_time,
                    "end_time": s.end_time,
                    "track_id": str(s.track_id)
                }
                for s in self.edit_operations.clipboard
            ]
        }

    def get_selection_info(self) -> Dict[str, Any]:
        """Get information about current selection."""
        return {
            "selected_segments": list(self.selected_segments),
            "selected_tracks": list(self.selected_tracks),
            "segment_count": len(self.selected_segments),
            "track_count": len(self.selected_tracks)
        }

    def validate_operation(self, operation: str, **kwargs) -> Dict[str, Any]:
        """Validate an edit operation before execution."""
        validation_result = {
            "valid": True,
            "errors": [],
            "warnings": []
        }

        if operation == "cut":
            cut_time = kwargs.get("cut_time", 0.0)
            if cut_time < 0:
                validation_result["valid"] = False
                validation_result["errors"].append("Cut time must be non-negative")

        elif operation == "paste":
            paste_time = kwargs.get("paste_time", 0.0)
            if paste_time < 0:
                validation_result["valid"] = False
                validation_result["errors"].append("Paste time must be non-negative")

        elif operation == "trim":
            trim_time = kwargs.get("trim_time", 0.0)
            if trim_time < 0:
                validation_result["valid"] = False
                validation_result["errors"].append("Trim time must be non-negative")

        return validation_result

    def get_operation_history(self) -> List[Dict[str, Any]]:
        """Get history of edit operations."""
        # This would return the actual operation history
        # For now, return empty list as placeholder
        return []

File: src/components/Timeline/EditControlsManager.ts
/**
 * EditControlsManager handles all timeline editing operations.
 *
 * This class provides comprehensive edit functionality including cut, copy, paste,
 * delete, trim, and split operations with full validation and state management.
 */

import { TimelineSegment, TimelineTrack, TimelineState } from '../../core/timeline/models';
import { TimelineStateManager } from '../../core/timeline/state';

export interface EditOperation {
  type: 'cut' | 'copy' | 'paste' | 'delete' | 'trim_start' | 'trim_end' | 'split';
  segments: TimelineSegment[];
  timestamp: number;
  previousState?: TimelineState;
  newState?: TimelineState;
}

export class EditControlsManager {
  private stateManager: TimelineStateManager;
  private currentState: TimelineState;
  private history: EditOperation[] = [];
  private historyIndex: number = -1;
  private maxHistorySize: number = 50;

  constructor(stateManager: TimelineStateManager) {
    this.stateManager = stateManager;
  }

  /**
   * Initialize the edit controls manager with the current timeline state.
   */
  initialize(state: TimelineState): void {
    this.currentState = JSON.parse(JSON.stringify(state)); // Deep copy
  }

  /**
   * Cut selected segments and add to clipboard.
   */
  async cutSegments(segments: TimelineSegment[]): Promise<TimelineSegment[]> {
    if (segments.length === 0) {
      throw new Error('No segments provided for cut operation');
    }

    // Validate segments exist in timeline
    this.validateSegments(segments);

    // Store current state for undo
    const previousState = JSON.parse(JSON.stringify(this.currentState));

    // Create copies of segments for clipboard
    const cutSegments = segments.map(segment => ({
      ...segment,
      id: `cut_${segment.id}_${Date.now()}`,
      selected: false
    }));

    // Remove segments from timeline
    this.currentState.tracks.forEach(track => {
      track.segments = track.segments.filter(segment =>
        !segments.some(s => s.id === segment.id)
      );
    });

    // Create and store operation
    const operation: EditOperation = {
      type: 'cut',
      segments: cutSegments,
      timestamp: Date.now(),
      previousState,
      newState: JSON.parse(JSON.stringify(this.currentState))
    };

    this.addToHistory(operation);

    // Update state manager
    await this.stateManager.updateState(this.currentState);

    return cutSegments;
  }

  /**
   * Copy selected segments to clipboard.
   */
  async copySegments(segments: TimelineSegment[]): Promise<TimelineSegment[]> {
    if (segments.length === 0) {
      throw new Error('No segments provided for copy operation');
    }

    // Validate segments exist in timeline
    this.validateSegments(segments);

    // Create copies of segments for clipboard
    const copiedSegments = segments.map(segment => ({
      ...segment,
      id: `copy_${segment.id}_${Date.now()}`,
      selected: false
    }));

    return copiedSegments;
  }

  /**
   * Paste segments from clipboard at specified time.
   */
  async pasteSegments(segments: TimelineSegment[], targetTime: number): Promise<TimelineSegment[]> {
    if (segments.length === 0) {
      throw new Error('No segments provided for paste operation');
    }

    if (targetTime < 0) {
      throw new Error('Invalid target time for paste operation');
    }

    // Store current state for undo
    const previousState = JSON.parse(JSON.stringify(this.currentState));

    // Calculate time offset for paste
    const timeOffset = targetTime - Math.min(...segments.map(s => s.startTime));

    // Create new segments with updated timing
    const pastedSegments = segments.map(segment => ({
      ...segment,
      id: `pasted_${segment.id}_${Date.now()}`,
      startTime: segment.startTime + timeOffset,
      endTime: segment.endTime + timeOffset,
      selected: false
    }));

    // Validate no overlaps with existing segments
    this.validateNoOverlaps(pastedSegments);

    // Add segments to appropriate tracks
    pastedSegments.forEach(segment => {
      const targetTrack = this.findOrCreateTrackForSegment(segment);
      targetTrack.segments.push(segment);
    });

    // Sort segments in each track
    this.currentState.tracks.forEach(track => {
      track.segments.sort((a, b) => a.startTime - b.startTime);
    });

    // Create and store operation
    const operation: EditOperation = {
      type: 'paste',
      segments: pastedSegments,
      timestamp: Date.now(),
      previousState,
      newState: JSON.parse(JSON.stringify(this.currentState))
    };

    this.addToHistory(operation);

    // Update state manager
    await this.stateManager.updateState(this.currentState);

    return pastedSegments;
  }

  /**
   * Delete selected segments.
   */
  async deleteSegments(segments: TimelineSegment[]): Promise<void> {
    if (segments.length === 0) {
      throw new Error('No segments provided for delete operation');
    }

    // Validate segments exist in timeline
    this.validateSegments(segments);

    // Store current state for undo
    const previousState = JSON.parse(JSON.stringify(this.currentState));

    // Remove segments from timeline
    this.currentState.tracks.forEach(track => {
      track.segments = track.segments.filter(segment =>
        !segments.some(s => s.id === segment.id)
      );
    });

    // Create and store operation
    const operation: EditOperation = {
      type: 'delete',
      segments: [...segments],
      timestamp: Date.now(),
      previousState,
      newState: JSON.parse(JSON.stringify(this.currentState))
    };

    this.addToHistory(operation);

    // Update state manager
    await this.stateManager.updateState(this.currentState);
  }

  /**
   * Trim start of a segment.
   */
  async trimStart(segment: TimelineSegment, trimTime: number): Promise<TimelineSegment> {
    if (trimTime <= segment.startTime || trimTime >= segment.endTime) {
      throw new Error('Invalid trim time for start trim operation');
    }

    // Store current state for undo
    const previousState = JSON.parse(JSON.stringify(this.currentState));

    // Find and update the segment
    let trimmedSegment: TimelineSegment | null = null;

    this.currentState.tracks.forEach(track => {
      const index = track.segments.findIndex(s => s.id === segment.id);
      if (index !== -1) {
        const originalSegment = track.segments[index];
        trimmedSegment = {
          ...originalSegment,
          id: `trimmed_${originalSegment.id}_${Date.now()}`,
          startTime: trimTime,
          duration: originalSegment.endTime - trimTime
        };

        // Replace the segment
        track.segments[index] = trimmedSegment;
      }
    });

    if (!trimmedSegment) {
      throw new Error('Segment not found for trim operation');
    }

    // Create and store operation
    const operation: EditOperation = {
      type: 'trim_start',
      segments: [trimmedSegment],
      timestamp: Date.now(),
      previousState,
      newState: JSON.parse(JSON.stringify(this.currentState))
    };

    this.addToHistory(operation);

    // Update state manager
    await this.stateManager.updateState(this.currentState);

    return trimmedSegment;
  }

  /**
   * Trim end of a segment.
   */
  async trimEnd(segment: TimelineSegment, trimTime: number): Promise<TimelineSegment> {
    if (trimTime <= segment.startTime || trimTime >= segment.endTime) {
      throw new Error('Invalid trim time for end trim operation');
    }

    // Store current state for undo
    const previousState = JSON.parse(JSON.stringify(this.currentState));

    // Find and update the segment
    let trimmedSegment: TimelineSegment | null = null;

    this.currentState.tracks.forEach(track => {
      const index = track.segments.findIndex(s => s.id === segment.id);
      if (index !== -1) {
        const originalSegment = track.segments[index];
        trimmedSegment = {
          ...originalSegment,
          id: `trimmed_${originalSegment.id}_${Date.now()}`,
          endTime: trimTime,
          duration: trimTime - originalSegment.startTime
        };

        // Replace the segment
        track.segments[index] = trimmedSegment;
      }
    });

    if (!trimmedSegment) {
      throw new Error('Segment not found for trim operation');
    }

    // Create and store operation
    const operation: EditOperation = {
      type: 'trim_end',
      segments: [trimmedSegment],
      timestamp: Date.now(),
      previousState,
      newState: JSON.parse(JSON.stringify(this.currentState))
    };

    this.addToHistory(operation);

    // Update state manager
    await this.stateManager.updateState(this.currentState);

    return trimmedSegment;
  }

  /**
   * Split a segment at specified time.
   */
  async splitSegment(segment: TimelineSegment, splitTime: number): Promise<TimelineSegment[]> {
    if (splitTime <= segment.startTime || splitTime >= segment.endTime) {
      throw new Error('Invalid split time for split operation');
    }

    // Store current state for undo
    const previousState = JSON.parse(JSON.stringify(this.currentState));

    // Find the segment and split it
    const splitSegments: TimelineSegment[] = [];

    this.currentState.tracks.forEach(track => {
      const index = track.segments.findIndex(s => s.id === segment.id);
      if (index !== -1) {
        const originalSegment = track.segments[index];

        // Create first segment
        const firstSegment: TimelineSegment = {
          ...originalSegment,
          id: `split1_${originalSegment.id}_${Date.now()}`,
          endTime: splitTime,
          duration: splitTime - originalSegment.startTime
        };

        // Create second segment
        const secondSegment: TimelineSegment = {
          ...originalSegment,
          id: `split2_${originalSegment.id}_${Date.now()}`,
          startTime: splitTime,
          duration: originalSegment.endTime - splitTime
        };

        // Replace original with split segments
        track.segments.splice(index, 1, firstSegment, secondSegment);
        splitSegments.push(firstSegment, secondSegment);
      }
    });

    if (splitSegments.length === 0) {
      throw new Error('Segment not found for split operation');
    }

    // Create and store operation
    const operation: EditOperation = {
      type: 'split',
      segments: splitSegments,
      timestamp: Date.now(),
      previousState,
      newState: JSON.parse(JSON.stringify(this.currentState))
    };

    this.addToHistory(operation);

    // Update state manager
    await this.stateManager.updateState(this.currentState);

    return splitSegments;
  }

  /**
   * Undo the last operation.
   */
  async undo(): Promise<TimelineSegment[]> {
    if (!this.canUndo()) {
      throw new Error('No operations to undo');
    }

    const operation = this.history[this.historyIndex];
    this.historyIndex--;

    if (operation.previousState) {
      this.currentState = JSON.parse(JSON.stringify(operation.previousState));
      await this.stateManager.updateState(this.currentState);
    }

    return operation.segments;
  }

  /**
   * Redo the last undone operation.
   */
  async redo(): Promise<TimelineSegment[]> {
    if (!this.canRedo()) {
      throw new Error('No operations to redo');
    }

    this.historyIndex++;
    const operation = this.history[this.historyIndex];

    if (operation.newState) {
      this.currentState = JSON.parse(JSON.stringify(operation.newState));
      await this.stateManager.updateState(this.currentState);
    }

    return operation.segments;
  }

  /**
   * Check if undo is available.
   */
  canUndo(): boolean {
    return this.historyIndex >= 0;
  }

  /**
   * Check if redo is available.
   */
  canRedo(): boolean {
    return this.historyIndex < this.history.length - 1;
  }

  /**
   * Get current state.
   */
  getCurrentState(): TimelineState {
    return JSON.parse(JSON.stringify(this.currentState));
  }

  /**
   * Validate that segments exist in timeline.
   */
  private validateSegments(segments: TimelineSegment[]): void {
    segments.forEach(segment => {
      const exists = this.currentState.tracks.some(track =>
        track.segments.some(s => s.id === segment.id)
      );

      if (!exists) {
        throw new Error(`Segment ${segment.id} not found in timeline`);
      }
    });
  }

  /**
   * Validate no overlaps with new segments.
   */
  private validateNoOverlaps(newSegments: TimelineSegment[]): void {
    this.currentState.tracks.forEach(track => {
      track.segments.forEach(existingSegment => {
        newSegments.forEach(newSegment => {
          if (this.segmentsOverlap(existingSegment, newSegment)) {
            throw new Error(`Segment overlap detected between ${existingSegment.id} and ${newSegment.id}`);
          }
        });
      });
    });
  }

  /**
   * Check if two segments overlap.
   */
  private segmentsOverlap(segment1: TimelineSegment, segment2: TimelineSegment): boolean {
    return segment1.startTime < segment2.endTime && segment2.startTime < segment1.endTime;
  }

  /**
   * Find or create appropriate track for a segment.
   */
  private findOrCreateTrackForSegment(segment: TimelineSegment): TimelineTrack {
    // For now, find the first available track or create a new one
    let targetTrack = this.currentState.tracks.find(track =>
      track.type === 'video' || track.type === 'audio'
    );

    if (!targetTrack) {
      targetTrack = {
        id: `track_${Date.now()}`,
        name: `Track ${this.currentState.tracks.length + 1}`,
        type: 'video',
        segments: [],
        muted: false,
        volume: 1.0,
        visible: true
      };
      this.currentState.tracks.push(targetTrack);
    }

    return targetTrack;
  }

  /**
   * Add operation to history.
   */
  private addToHistory(operation: EditOperation): void {
    // Remove any operations after current index (when new operation after undo)
    if (this.historyIndex < this.history.length - 1) {
      this.history = this.history.slice(0, this.historyIndex + 1);
    }

    // Add new operation
    this.history.push(operation);
    this.historyIndex++;

    // Limit history size
    if (this.history.length > this.maxHistorySize) {
      this.history = this.history.slice(-this.maxHistorySize);
      this.historyIndex = Math.min(this.historyIndex, this.maxHistorySize - 1);
    }
  }
}

export default EditControlsManager;